event/consumer.go

```diff
@@ -9,61 +12,99 @@ import (
 // MessageConsumer provides a generic interface for consuming []byte messages (from Kafka)
 type MessageConsumer interface {
 	Incoming() chan kafka.Message
-	Closer() chan bool
 }
 
 // Handler represents a handler for processing a batch of events.
 type Handler interface {
 	Handle(events []*ObservationExtracted) error
 }
 
+type Consumer struct {
+	closing chan bool
+	closed  chan bool
+}
+
+// NewConsumerGroup returns a new consumer group using default configuration.
+func NewConsumer() *Consumer {
+
+	c := Consumer{
+		closing: make(chan bool),
+		closed:  make(chan bool),
+	}
+
+	return &c
+}
+
 // Consume convert them to event instances, and pass the event to the provided handler.
-func Consume(messageConsumer MessageConsumer,
+func (consumer *Consumer) Consume(messageConsumer MessageConsumer,
 	batchSize int,
 	handler Handler,
 	batchWaitTime time.Duration,
-	exit chan struct{}) {
+	error chan error) {
 
-	batch := NewBatch(batchSize)
+	go func() {
+		defer close(consumer.closed)
 
-	// Wait a batch full of messages.
-	// If we do not get any messages for a time, just process the messages already in the batch.
-	for {
-		select {
-		case msg := <-messageConsumer.Incoming():
+		batch := NewBatch(batchSize)
 
-			AddMessageToBatch(batch, msg, handler, exit)
+		// Wait a batch full of messages.
+		// If we do not get any messages for a time, just process the messages already in the batch.
+		for {
+			select {
+			case msg := <-messageConsumer.Incoming():
 
-		case <-time.After(batchWaitTime):
+				AddMessageToBatch(batch, msg, handler, error)
 
-			if batch.IsEmpty() {
-				continue
-			}
+			case <-time.After(batchWaitTime):
 
-			log.Debug("batch wait time reached. proceeding with batch", log.Data{"batchsize": batch.Size()})
-			ProcessBatch(handler, batch, exit)
+				if batch.IsEmpty() {
+					continue
+				}
 
-		case <-exit:
-			return
+				log.Debug("batch wait time reached. proceeding with batch", log.Data{"batchsize": batch.Size()})
+				ProcessBatch(handler, batch, error)
+
+			case <-consumer.closing:
```

Comment:
Possibly add a log statement here to indicate the consumer loop has been closed.

Reply:
Done